docs/proposals/20230407-managed-k8s-capi-contract-changes.md

```diff
@@ -0,0 +1,380 @@
+
+This flow would require making changes to CAPI controllers so that there is an option to report the `ControlPlaneEndpoint` via the `<Infra>ControlPlane` as an alternative to coming from the `<Infra>Cluster`.
+
+Using CAPG as an example:
+
+```go
+type GCPManagedControlPlaneSpec struct {
+  // AddonsConfig defines the addons to enable with the GKE cluster.
+  // +optional
+  AddonsConfig *AddonsConfig `json:"addonsConfig,omitempty"`
+
+  // Logging contains the logging configuration for the GKE cluster.
+  // +optional
+  Logging *ControlPlaneLoggingSpec `json:"logging,omitempty"`
+
+  // EnableKubernetesAlpha will indicate the kubernetes alpha features are enabled
+  // +optional
+  EnableKubernetesAlpha bool
+
+  // ControlPlaneEndpoint represents the endpoint used to communicate with the control plane.
+  // +optional
+  ControlPlaneEndpoint clusterv1.APIEndpoint `json:"controlPlaneEndpoint"`
+
+  ...
+}
+
+
+type GCPManagedClusterSpec struct {
+  // Project is the name of the project to deploy the cluster to.
+  Project string `json:"project"`
+
+  // The GCP Region the cluster lives in.
+  Region string `json:"region"`
+
+  // NetworkSpec encapsulates all things related to the GCP network.
+  // +optional
+  Network NetworkSpec `json:"network"`
+
+  // FailureDomains is an optional field which is used to assign selected availability zones to a cluster
+  // FailureDomains if empty, defaults to all the zones in the selected region and if specified would override
+  // the default zones.
+  // +optional
+  FailureDomains []string `json:"failureDomains,omitempty"`
+
+  ...
+}
+```
+
+**Pros**
+
+- Simplifies provider implementation when reporting `ControlPlaneEndpoint`
+- Clearer separation between the lifecycle management of the general cloud infrastructure required for the cluster and the actual managed control plane (GKE in this example)
+- Follows the original intentions of an "infrastructure" and "control-plane" provider
+- Enables removal/addition of properties for a Managed Kubernetes cluster that may be different from a self-managed Kubernetes cluster
+- Works with ClusterClass
+
+**Cons**
+
+- Requires changes upstream to CAPI controllers to support the change of reporting `ControlPlaneEndpoint`
+- Duplication of API definitions between self-managed and managed `<Infra>Cluster` definitions and related controllers
+- Users need to be aware of when to use the unmanaged or managed `<Infra>Cluster` definitions.
+
+#### Flow 2: Change CAPI to make `<Infra>Cluster` optional
```

Comment:
Might have missed this, but what use cases does this flow solve?

Reply:
This solves for a Managed Kubernetes service solution that obscures cloud infra entirely from the user and exposes only control plane configuration (apiserver endpoint, k8s version as simple example. As distinct from a Managed Kubernetes service solution that gave users lots of flexibility to define (or use pre-existing) infra like network, resource management, etc that our current best-practice docs say "put this stuff in your <Infra>ManagedCluster resource when implementing managed k8s in your CAPI provider".

Reply:
Was this asked by other provider implementers?

Reply:
There are AKS scenarios which agree with this description.